lib/api/api-fetch/response.js

```diff
@@ -1,141 +1,367 @@
 
+// https://fetch.spec.whatwg.org/#response-class
 class Response {
-  constructor ({
-    type,
-    url,
-    body,
-    statusCode,
-    headers,
-    context
-  }) {
-    this[kType] = type || 'default'
-    this[kStatus] = statusCode || 0
-    this[kStatusText] = STATUS_CODES[statusCode] || ''
-    this[kUrlList] = Array.isArray(url) ? url : (url ? [url] : [])
-    this[kHeaders] = headers || new Headers()
-    this[kBody] = body || null
-
-    if (context && context.history) {
-      this[kUrlList].push(...context.history)
+  // Creates network error Response.
+  static error () {
+    // The static error() method steps are to return the result of creating a
+    // Response object, given a new network error, "immutable", and this’s
+    // relevant Realm.
+    // TODO: relevant Realm?
+    const responseObject = new Response()
+    responseObject[kState] = makeNetworkError()
+    responseObject[kHeaders][kHeadersList] = responseObject[kState].headersList
+    responseObject[kHeaders][kGuard] = 'immutable'
+    return responseObject
+  }
+
+  // Creates a redirect Response that redirects to url with status status.
+  static redirect (url, status = 302) {
+    // 1. Let parsedURL be the result of parsing url with current settings
+    // object’s API base URL.
+    // 2. If parsedURL is failure, then throw a TypeError.
+    // TODO: base-URL?
+    let parsedURL
+    try {
+      parsedURL = new URL(url)
+    } catch (err) {
+      const error = new TypeError()
+      error.cause = err
+      throw error
+    }
+
+    // 3. If status is not a redirect status, then throw a RangeError.
+    if (!redirectStatus.includes(status)) {
+      throw new RangeError()
+    }
+
+    // 4. Let responseObject be the result of creating a Response object,
+    // given a new response, "immutable", and this’s relevant Realm.
+    // TODO: relevant Realm?
+    const responseObject = new Response()
+    responseObject[kHeaders][kGuard] = 'immutable'
+
+    // 5. Set responseObject’s response’s status to status.
+    responseObject[kState].status = status
+
+    // 6. Let value be parsedURL, serialized and isomorphic encoded.
+    // TODO: isomorphic encoded?
+    const value = parsedURL.toString()
+
+    // 7. Append `Location`/value to responseObject’s response’s header list.
+    responseObject[kState].headersList.push('location', value)
+
+    // 8. Return responseObject.
+    return responseObject
+  }
+
+  // https://fetch.spec.whatwg.org/#dom-response
+  constructor (body = null, init = {}) {
+    // 1. If init["status"] is not in the range 200 to 599, inclusive, then
+    // throw a RangeError.
+    if ('status' in init) {
+      if (!Number.isFinite(init.status)) {
+        throw new TypeError()
+      }
+
+      if (init.status < 200 || init.status > 599) {
+        throw new RangeError()
+      }
+    }
+
+    if ('statusText' in init) {
+      if (typeof init.statusText !== 'string') {
+        throw new TypeError()
```

Comment:
lol this is actually the one where browsers fail:

```js
new Response(null, { statusText: 0 })
```

both Chrome and Firefox pass. A more descriptive error message would be nice 👍🏼

Reply:
What do they do? Just stringify? Maybe that is actually more correct.